### PR TITLE
Heterogeneous Betas

### PR DESCRIPTION
**Pull request by jdebacker**
This PR updates the model to allow for heterogenous beta, with the ability to vary beta over the lifetime income group, j.

As shown in [Carroll, Slacalek, Tokuoka, and White (*Quantitative Economics*, 2017)](http://www.econ2.jhu.edu/people/ccarroll/papers/cstwMPC/), heterogenous values fo the discount factor are important for accounting for the distribution of wealth and more realistic marginal propensities to consume.

### CODE REVIEW

File: ogusa/default_parameters.json [hunk 1]
```diff
@@ -185,11 +185,20 @@
...
         "section_1": "Household Parameters",
         "section_2": "Behavioral Assumptions",
-        "notes": "Default value from Carroll (JME, 2009).",
+        "notes": "Default value from Carroll (JME, 2009).  Allows for each type to have a different value.",
         "type": "float",
+        "number_dims": 1,
         "value": [
             {
-                "value": 0.96
+                "value": [
```
Comment (MaxGhenis) [inline ogusa/default_parameters.json:193]: Did you have a paper in mind for assigning different defaults by j? Or is the idea to calibrate them as part of OG-USA to better match wealth and MPC distributions?

I'm doing separate research that involves time preference across the income distribution, so this PR caught my eye. I didn't see it in Carroll (2017) but only gave it a cursory look.

**Review by MaxGhenis — COMMENTED** [review]
1 q

**Comment by jdebacker** [comment]
@MaxGhenis This PR is just going to introduce the infrastructure for heterogenous betas.  The calibration can be handled elsewhere.

But, to your questions, Carroll et al (2017) do estimate the distribution of betas (see Section 3.5).  Their context is slightly different (with no heterogeneity in permanent income earning ability as we have here), but my thought is to take a similar empirical approach (matching moments in the distribution of wealth) as they do, which is what we have done in the past to calibrate the chi_b values.

**Comment by jdebacker** [comment]
To compare results, here are the SS lifecycle profiles of savings with a constant beta:

SSLifecycleProfile_Save_Baseline_constant_beta

And with heterogenous betas (calibrated to Carroll et al. (2017)) values:

SSLifecycleProfile_Save_Baseline_hetero_beta

cc @rickecon

**Comment by jdebacker** [comment]
FYI, this PR passes the full set of unit tests run locally.

**Comment by jdebacker** [comment]
@rickecon This PR is ready for your review

**Comment by rickecon** [comment]
@jdebacker Two quick questions.
1. Why is your lifecycle profile of savings with constant betas figure above so different from the corresponding figure in [your comment](https://github.com/PSLmodels/OG-USA/pull/652#issuecomment-700361537) in PR #652?
2. Why are all of the savings levels in your heterogeneous betas figure above higher than the corresponding constant betas levels? Are the Carroll (2017) calibrated betas all higher than our constant beta?

If the figure in [your comment](https://github.com/PSLmodels/OG-USA/pull/652#issuecomment-700361537) in PR #652 is not correct and the constant betas figure above is correct, then the heterogeneous betas don't seem to make that big a difference in the distribution of savings by age and lifetime income group.

**Comment by jdebacker** [comment]
> Why is your lifecycle profile of savings with constant betas figure above so different from the corresponding figure in your comment in PR #652?

This is because picture in PR #652 has a substantially different calibration of the income processes.  In particular, the top group in that plot is the top 0.1% of earners, who earn about 40x more per hour than the top 1% which is represented in the plots above.

> Why are all of the savings levels in your heterogeneous betas figure above higher than the corresponding constant betas levels? Are the Carroll (2017) calibrated betas all higher than our constant beta?

Carroll et al (2017) has a higher mean beta the the default in OG-USA (0.9659 in Carroll et al vs 0.96 in OG-USA).  The betas with the Carroll et al (2017) calibration are `beta = array([0.93598571, 0.94595714, 0.95592857, 0.9659, 0.97587143, 0.98584286, 0.99581429])`.  Income groups above `j=1` have higher betas than in the OG-USA calibration and so higher savings levels.

> If the figure in your comment in PR #652 is not correct and the constant betas figure above is correct, then the heterogeneous betas don't seem to make that big a difference in the distribution of savings by age and lifetime income group.

The difference of course depends on the calibration, which is beyond the scope of this PR.  But as you note above, the betas make a big difference in the level and distribution of savings (e.g., savings of `j=7` is about 2x as much with the Carroll et al calibration).  This is pretty substantial and confirms what Carroll et al find - that heterogenous betas can be very important for matching the distribution of wealth.  The lifecycle profiles are similarly shaped, but those shapes are largely driven by the lifecycle earnings profile and the bequest motive.

This PR is intended to allow one to work with heterogenous betas in OG-USA.  A separate calibration exercise should be used to set values to these betas, jointly with values of other parameters that affect savings and labor supply decisions.  This PR will allow one the ability to do such an exercise.

**Comment by rickecon** [comment]
@jdebacker I reviewed the code and this looks great. I was pleasantly surprised how little needed to change in order to add this feature. I have submitted [a PR to your branch](https://github.com/jdebacker/dynamic/pull/17) that has updates to the documentation to add to your PR. As soon as you merge that and it passes CI tests, I'll merge this PR.

**Pull request merged**